### PR TITLE
caprine-bin: init at 2.55.2

### LINKED ISSUE #90375: caprine
**Project description**
Caprine is an unofficial and privacy-focused Facebook Messenger app with many useful features.

**Metadata**
* homepage URL: <https://sindresorhus.com/caprine/>
* source URL: <https://github.com/sindresorhus/caprine>
* license: mit
* platforms: linux, darwin

### PR DESCRIPTION
**Pull request by ShamrockLee**
###### Motivation for this change
[Caprine](https://github.com/sindresorhus/caprine) is a third-party Facebook messenger client with a beautiful UI and features to enhance the privacy of users.
As an NodeJS-Electron application, the build-from-source attempt suffer from NPM trying to download Electron at build time, and the binary build can serve as a workaround.

Update:
Closes #90375 

###### Known issue

~~Crash when doing things involving "open files with file manager", complaining:~~
```
WARNING: Kernel has no file descriptor comparison support: Function not implemented

(caprine:22216): GLib-GIO-ERROR **: 05:53:43.885: No GSettings schemas are installed on the system
Trace/breakpoint trap (core dumped)

```
~~Tried adding some Gnome-related packages to extraPagkages, but it doesn't solve the problem. It has been mentioned [on NixOS Discourse](https://discourse.nixos.org/t/failure-appimage-run-with-gsetting/4403).~~

~~This is common on AppImages, and is also happen when executing using appimage-run.~~

~~Nevertheless, drag-and-drop of files still work, which can serve as a workaround to the problem above.~~

__Solved__ after referencing [dtzWill's work on minetime](https://github.com/NixOS/nixpkgs/blob/48feb84a9d91ac2601ac33e660d9f6c76f476a1e/pkgs/applications/office/minetime/default.nix#L31-L34) and add
```Nix
appimageTools.wrapType2 {
# ...
  profile = ''
    export LC_ALL=C.UTF-8
    export XDG_DATA_DIRS=${gsettings-desktop-schemas}/share/gsettings-schemas/${gsettings-desktop-schemas.name}:${gtk3}/share/gsettings-schemas/${gtk3.name}:$XDG_DATA_DIRS
  '';
# ...
}
```

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`) (The binary build from AppImage works. I don't have Mac with me, so I cannot test the dmg part.)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ShamrockLee** [comment]
@GrahamcOfBorg build caprine-bin

**Comment by ShamrockLee** [comment]
Solve the GSetting I/O issue by adding
```Nix
appimageTools.wrapType2 {
# ...
  profile = ''
    export LC_ALL=C.UTF-8
    export XDG_DATA_DIRS=${gsettings-desktop-schemas}/share/gsettings-schemas/${gsettings-desktop-schemas.name}:${gtk3}/share/gsettings-schemas/${gtk3.name}:$XDG_DATA_DIRS
  '';
# ...
}
```

**Comment by ShamrockLee** [comment]
Apply #108938 and use `lib` instead of `stdenv.lib`.

File: pkgs/applications/networking/instant-messengers/caprine-bin/dmg.nix [hunk 1]
```diff
@@ -0,0 +1,31 @@
+{ stdenv, lib, fetchurl, undmg, version }:
+
+let
+  pname = "caprine";
+  nameCamel = "Caprine";
+  nameSource = "${nameCamel}-${version}.dmg";
+  nameApp = nameCamel + ".app";
+  src = fetchurl {
+    url =
+      "https://github.com/sindresorhus/caprine/releases/download/v${version}/${nameSource}";
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/caprine-bin/dmg.nix:10]:
```suggestion
    url = "https://github.com/sindresorhus/caprine/releases/download/v${version}/${nameSource}";
```

File: pkgs/applications/networking/instant-messengers/caprine-bin/dmg.nix [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    name = nameSource;
+    sha256 = "1043g1qcykdxd4dp7pcz6bp3wg1ldmimvnlv39glmxwbblb6i9yi";
+  };
+in stdenv.mkDerivation {
+
+  inherit pname version src;
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/caprine-bin/dmg.nix:16]:
```suggestion
in stdenv.mkDerivation {
  inherit pname version src;
```

File: pkgs/applications/networking/instant-messengers/caprine-bin/appimage.nix [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    mkdir $out/share/applications
+    cp ${extracted}/${nameExecutable}.desktop $out/share/applications
+    substituteInPlace $out/share/applications/${nameExecutable}.desktop \
+        --replace AppRun ${nameExecutable}
+  '';
+
+  meta.platforms = with lib.platforms;
+    (lib.lists.intersectLists x86_64 (
+      linux ++ freebsd ++ netbsd ++ openbsd
+    ));
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/caprine-bin/appimage.nix:45]:
```suggestion
  meta.platforms = with lib.platforms [ x86_64-linux ];
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/applications/networking/instant-messengers/caprine-bin/appimage.nix (hunk 3 above)
Comment (ShamrockLee) [reply pkgs/applications/networking/instant-messengers/caprine-bin/appimage.nix]: It should be
```Nix
  meta.platforms = with lib.platforms; [ "x86_64-linux" ];
```

Thank you for the suggestions!

**Comment by ShamrockLee** [comment]
Rebased onto the recent master branch and upgrade to 2.52.2

**Comment by ShamrockLee** [comment]
I have personally been using this for moths, and I think it's ready to get merged.

**Comment by ShamrockLee** [comment]
@ofborg build caprine-bin

**Comment by austinbutler** [comment]
Result of `nixpkgs-review pr 108400` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>caprine-bin</li>
  </ul>
</details>

**Comment by austinbutler** [comment]
Works for me on Linux and builds on macOS (although not tested there).

**Comment by ShamrockLee** [comment]
Just did some renaming (`appimage-x86_64` -> `x86_64-appimage`, `build-appimage.nix` -> `build-from-appimage.nix` ). The result hash should not be affected.

File: pkgs/applications/networking/instant-messengers/caprine-bin/build-from-dmg.nix [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+{ pname
+, version
+, sha256 ? lib.fakeSha256
+, metaCommon ? { }
+}:
+
+let
+  nameApp = "Caprine.app";
+in
+stdenvNoCC.mkDerivation {
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/caprine-bin/build-from-dmg.nix:16]:
```suggestion
stdenvNoCC.mkDerivation {
```

File: pkgs/applications/networking/instant-messengers/caprine-bin/build-from-dmg.nix [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    inherit sha256;
+  };
+
+  sourceRoot = nameApp;
+
+  nativeBuildInputs = [ undmg ];
+
+  installPhase = ''
+    mkdir -p $out/Applications/${nameApp}
+    cp -R . $out/Applications/${nameApp}
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/caprine-bin/build-from-dmg.nix:32]:
```suggestion
    mkdir -p $out/Applications/Caprine.app
    cp -R . $out/Applications/Caprine.app
```

File: pkgs/applications/networking/instant-messengers/caprine-bin/build-from-appimage.nix [hunk 6]
```diff
@@ -0,0 +1,55 @@
+{ lib
+, fetchurl
+, appimageTools
+, gtk3
+, gsettings-desktop-schemas
+, lndir ? xorg.lndir
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/caprine-bin/build-from-appimage.nix:6]:
```suggestion
```

File: pkgs/applications/networking/instant-messengers/caprine-bin/build-from-appimage.nix [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    export XDG_DATA_DIRS=${gsettings-desktop-schemas}/share/gsettings-schemas/${gsettings-desktop-schemas.name}:${gtk3}/share/gsettings-schemas/${gtk3.name}:$XDG_DATA_DIRS
+  '';
+
+  extraPkgs = pkgs: appimageTools.defaultFhsEnvArgs.multiPkgs pkgs;
+
+  extraInstallCommands = ''
+    mv $out/bin/{${name},${nameExecutable}}
+
+    mkdir -p $out/share
+    "${lndir}/bin/lndir" -silent "${extracted}/usr/share" "$out/share"
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/caprine-bin/build-from-appimage.nix:41]:
```suggestion
    "${xorg.lndir}/bin/lndir" -silent "${extracted}/usr/share" "$out/share"
```

File: pkgs/top-level/all-packages.nix [hunk 8]
```diff
@@ -24484,6 +24484,8 @@ with pkgs;
 
   canto-daemon = callPackage ../applications/networking/feedreaders/canto-daemon { };
 
+  caprine-bin = callPackage ../applications/networking/instant-messengers/caprine-bin/default.nix { };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:24487]:
```suggestion
  caprine-bin = callPackage ../applications/networking/instant-messengers/caprine-bin { };
```

File: pkgs/applications/networking/instant-messengers/caprine-bin/build-from-dmg.nix [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+, sha256 ? lib.fakeSha256
+, metaCommon ? { }
+}:
+
+stdenvNoCC.mkDerivation {
+  inherit pname version;
+
+  src = fetchurl {
+    url =
+      "https://github.com/sindresorhus/caprine/releases/download/v${version}/Caprine-${version}.dmg";
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/caprine-bin/build-from-dmg.nix:18]:
```suggestion
    url = "https://github.com/sindresorhus/caprine/releases/download/v${version}/Caprine-${version}.dmg";
```

File: pkgs/applications/networking/instant-messengers/caprine-bin/build-from-dmg.nix [hunk 10]
```diff
@@ -0,0 +1,38 @@
+{ stdenvNoCC
+, lib
+, fetchurl
+, undmg
+}:
+
+{ pname
+, version
+, sha256 ? lib.fakeSha256
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/caprine-bin/build-from-dmg.nix:9]:
```suggestion
, sha256
```
sha256 should always be supplied and can be overwritten for testing purposes to lib.fakeSha256.

File: pkgs/applications/networking/instant-messengers/caprine-bin/build-from-appimage.nix [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+, sha256 ? lib.fakeSha256
+, metaCommon ? { }
+}:
+
+let
+  name = "${pname}-${version}";
+  nameExecutable = "caprine";
+  src = fetchurl {
+    url =
+      "https://github.com/sindresorhus/caprine/releases/download/v${version}/Caprine-${version}.AppImage";
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/caprine-bin/build-from-appimage.nix:20]:
```suggestion
    url = "https://github.com/sindresorhus/caprine/releases/download/v${version}/Caprine-${version}.AppImage";
```

File: pkgs/applications/networking/instant-messengers/caprine-bin/build-from-appimage.nix [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+{ pname
+, version
+, sha256 ? lib.fakeSha256
+, metaCommon ? { }
+}:
+
+let
+  name = "${pname}-${version}";
+  nameExecutable = "caprine";
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/caprine-bin/build-from-appimage.nix:17]:
```suggestion
```
That is identical to pname.

File: pkgs/applications/networking/instant-messengers/caprine-bin/build-from-appimage.nix [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+    ln -s ${extracted}/caprine.png $out/share/icons/caprine.png
+    mkdir $out/share/applications
+    cp ${extracted}/caprine.desktop $out/share/applications/
+    substituteInPlace $out/share/applications/caprine.desktop \
+        --replace AppRun ${nameExecutable}
+  '';
+
+  meta = metaCommon // {
+    platforms = [ "x86_64-linux" ];
+    mainProgram = nameExecutable;
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/caprine-bin/build-from-appimage.nix:50]:
```suggestion
    mv $out/bin/{${name},${pname}}

    mkdir -p $out/share
    "${xorg.lndir}/bin/lndir" -silent "${extracted}/usr/share" "$out/share"
    ln -s ${extracted}/caprine.png $out/share/icons/caprine.png
    mkdir $out/share/applications
    cp ${extracted}/caprine.desktop $out/share/applications/
    substituteInPlace $out/share/applications/caprine.desktop \
        --replace AppRun ${pname}
  '';

  meta = metaCommon // {
    platforms = [ "x86_64-linux" ];
    mainProgram = pname;
```

**Comment by austinbutler** [comment]
This worked for me up until recently. If I checkout this branch and merge latest master into it it works again for me. @SuperSandro2000 seems all your requests were satisfied, mind looking again?

**Comment by emilytrau** [comment]
🥳 nice!

**Pull request merged**